### PR TITLE
reset methods

### PR DESCRIPTION
**Pull request by yespire**

### CODE REVIEW

File: src/Common/Select.php [hunk 1]
```diff
@@ -751,6 +751,34 @@ protected function reset()
     }
 
     /**
+     * if user needs to reset
+     */
+    public function resetCols()
+    {
+        $this->cols = [];
```
Comment (harikt) [inline src/Common/Select.php]: php 5.3 , don't allow short array. So the tests is failing.

**Comment by harikt** [comment]
Do we need to return `$this` for method chaining ? Looking at other methods.

Additionally we need to add a few tests.

**Comment by yespire** [comment]
this package seems assuming no chaining in most method

**Comment by staabm** [comment]
Whats the use-case? Resetting objects is an anti pattern..

**Comment by yespire** [comment]
actually there don't seem to be a consistent return vs non-return $this

some methods return $this --- ie: setPaging()
some methods doesn't return $this --- ie: reset()

**Comment by yespire** [comment]
@staabm 

could you please elaborate "unsetting object is an anti-pattern" ?

#84 explained a use case

but here is another one:

the packages calcFoundRows() doesn't have a paired up method and property to capture "SELECT FOUND_ROWS()", even if it is added, some people may prefer to not use that and use SELECT COUNT(*)
//
// now imagine we need to get a list of records, then get the potential record count, the would be like:
//
$recordset = $query->fetchRecordSet();

$query->resetCols()
$query->resetOrderBy()
$query->resetLimit()
$query->resetOffset()
$query->cols([ "count(*)" ])
$listing_total = $query->fetchValue();

**Comment by staabm** [comment]
You should add the counterpart instead of relying on impl. Details and partly reset the object .

Noone can guarantee that the query objects will behave the way you need it across versions

**Comment by yespire** [comment]
@staabm 

That example is not implying don't add calcFoundRows()'s getter

IF there is no edit/unset/reset methods, the query building process become "add-only"
Can you explain why edit / unset / reset part of query object is bad ?

**Comment by staabm** [comment]
Objects can depend on several properties and additional properties could be added in the future. In case you rely on resettig certain properties your logic can/will start failing when the object your are resetting gets new properties in the future (which can happen even in minor releases)

**Comment by yespire** [comment]
@staabm

What you said is true "in general", However I don't see that applies here.

1.
Cols/OrderBy/Limit/Offset of Query/Select should be pretty stable, which probably won't change in near future. Maybe you are referring to something else ?

2.
FOUND_ROWS() has its own issues, implementing counterpart methods for Mysql/Select :: calcFoundRows() may not be helpful, there is an discussion regarding this:
https://github.com/atlasphp/Atlas.Orm/issues/5

3.
Only parent query builder object is depending on these "Cols, OrderBy, Where, etc". when we use query builder, we normally go through 2 steps: constructing phase, which business logics are checked to add / remove / modify the query builder states/properties, when all logics are checked, it builds a statement. The query builder only impact other objects via final statement. Ability to reset / modify Cols, OrderBy, Where is necessary feature for a query builder.

**Comment by pavarnos** [comment]
Could we work around the issues @staabm raises by adding a selectTotalItemCount() method which internally resets the correct parts of the query, adds a COUNT(*) or SQL_CALC_FOUND_ROWS? This would hide the implementation details he is concerned about and make pagination easy...

**Comment by harikt** [comment]
@pavarnos I like the idea.

**Comment by gauthier** [comment]
This is huge! I just pushed a very similar PR, that I had in mind for a while, but I did not check someone already did!

Sorry for the duplicate PR :(

**Pull request merged**